File: androidApp/src/main/java/com/cocktailcraft/android/viewmodel/OrderViewModelSKIE.kt
package com.cocktailcraft.android.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.cocktailcraft.domain.model.CocktailCartItem
import com.cocktailcraft.domain.model.Order
import com.cocktailcraft.domain.model.OrderStatus
import com.cocktailcraft.viewmodel.SharedOrderViewModel
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

/**
 * Android ViewModel wrapper for SharedOrderViewModel using SKIE.
 * This replaces the old OrderViewModel with a thin wrapper around the shared implementation.
 * 
 * Key SKIE benefits:
 * - StateFlows work directly without conversion
 * - Suspend functions work with Android coroutines
 * - Shared business logic reduces code duplication
 * - Type safety preserved across platforms
 */
class OrderViewModelSKIE : ViewModel(), KoinComponent {
    
    // Inject the shared ViewModel
    private val sharedViewModel: SharedOrderViewModel by inject()
    
    // Expose StateFlows from shared ViewModel (SKIE handles the interop)
    val orders: StateFlow<List<Order>> = sharedViewModel.orders
    val currentOrder: StateFlow<Order?> = sharedViewModel.currentOrder
    val orderCount: StateFlow<Int> = sharedViewModel.orderCount
    val isPlacingOrder: StateFlow<Boolean> = sharedViewModel.isPlacingOrder
    val totalSpent: StateFlow<Double> = sharedViewModel.totalSpent
    
    // Expose loading and error from shared base class
    val loadingState: StateFlow<Boolean> = sharedViewModel.isLoading
    val errorState = sharedViewModel.error
    val errorEventState = sharedViewModel.errorEvent
    
    // Computed properties
    val hasOrders: Boolean
        get() = sharedViewModel.hasOrders
    
    val isEmpty: Boolean
        get() = sharedViewModel.isEmpty
    
    // MARK: - Public Methods (SKIE converts these to proper suspend functions)
    
    /**
     * Load orders using SKIE async interop
     */
    fun loadOrders() {
        viewModelScope.launch {
            sharedViewModel.loadOrders()
        }
    }
    
    /**
     * Place a new order using SKIE async interop
     */
    fun placeOrder(cartItems: List<CocktailCartItem>, totalPrice: Double) {
        viewModelScope.launch {
            sharedViewModel.placeOrder(cartItems, totalPrice)
        }
    }
    
    /**
     * Get order by ID using SKIE async interop
     */
    fun getOrderById(orderId: String) {
        viewModelScope.launch {
            sharedViewModel.getOrderById(orderId)
        }
    }
    
    /**
     * Update order status using SKIE async interop
     */
    fun updateOrderStatus(orderId: String, status: String) {
        viewModelScope.launch {
            sharedViewModel.updateOrderStatus(orderId, status)
        }
    }
    
    /**
     * Cancel order using SKIE async interop
     */
    fun cancelOrder(orderId: String) {
        viewModelScope.launch {
            sharedViewModel.cancelOrder(orderId)
        }
    }
    
    /**
     * Reorder items using SKIE async interop
     */
    fun reorderItems(orderId: String) {
        viewModelScope.launch {
            sharedViewModel.reorderItems(orderId)
        }
    }
    
    // MARK: - Synchronous Methods (direct delegation)
    
    /**
     * Get orders by status
     */
    fun getOrdersByStatus(status: String): List<Order> {
        return sharedViewModel.getOrdersByStatus(status)
    }
    
    /**
     * Get total amount spent
     */
    fun getTotalSpent(): Double {
        return sharedViewModel.getTotalSpent()
    }
    
    /**
     * Get order history with limit
     */
    fun getOrderHistory(limit: Int = 10): List<Order> {
        return sharedViewModel.getOrderHistory(limit)
    }
    
    /**
     * Get recent orders
     */
    fun getRecentOrders(): List<Order> {
        return sharedViewModel.getRecentOrders()
    }
    
    /**
     * Get orders by date range
     */
    fun getOrdersByDateRange(startDate: String, endDate: String): List<Order> {
        return sharedViewModel.getOrdersByDateRange(startDate, endDate)
    }
    
    /**
     * Get order statistics
     */
    fun getOrderStatistics(): Map<String, Any> {
        return sharedViewModel.getOrderStatistics()
    }
    
    /**
     * Check if order can be cancelled
     */
    fun canCancelOrder(orderId: String): Boolean {
        return sharedViewModel.canCancelOrder(orderId)
    }
    
    /**
     * Get order status display text
     */
    fun getOrderStatusDisplay(status: String): String {
        return sharedViewModel.getOrderStatusDisplay(status)
    }
    
    /**
     * Get estimated delivery time for an order
     */
    fun getEstimatedDeliveryTime(orderId: String): String {
        return sharedViewModel.getEstimatedDeliveryTime(orderId)
    }
    
    /**
     * Clear the current error
     */
    fun clearErrorState() {
        sharedViewModel.clearError()
    }
    
    /**
     * Refresh orders data
     */
    fun refresh() {
        sharedViewModel.refresh()
    }
    
    /**
     * Clean up when ViewModel is cleared
     */
    override fun onCleared() {
        super.onCleared()
        sharedViewModel.onCleared()
    }
}

File: iosApp/CocktailCraft/ViewModels/OrderViewModelSKIE.swift
import SwiftUI
import shared

/**
 * iOS ViewModel wrapper for SharedOrderViewModel using pure SKIE integration.
 * State/error mirroring and observation-task lifecycle live in
 * SharedViewModelWrapper.
 */
final class OrderViewModelSKIE: SharedViewModelWrapper<OrderUiState> {

    // Computed properties
    var hasOrders: Bool {
        !state.orders.isEmpty
    }

    var isEmpty: Bool {
        state.orders.isEmpty
    }

    var recentOrders: [Order] {
        sharedViewModel.getRecentOrders()
    }

    var orderStatistics: OrderStatistics {
        sharedViewModel.getOrderStatistics()
    }

    // Shared ViewModel instance
    private let sharedViewModel: SharedOrderViewModel

    init() {
        let viewModel = getSharedKoinHelper().getSharedOrderViewModel()
        self.sharedViewModel = viewModel
        super.init(uiState: viewModel.uiState, errorFlow: viewModel.error)
    }

    // No deinit: the base class cancels observation. Wraps a Koin `single`
    // whose coroutine scope must survive any one wrapper — never onCleared().

    // MARK: - Public Methods (using SKIE async/await)

    func loadOrders() async {
        await run { try await sharedViewModel.loadOrders() }
    }

    func placeOrder(cartItems: [CocktailCartItem], totalPrice: Double) async -> Bool {
        return await run(fallback: false) { try await sharedViewModel.placeOrder(cartItems: cartItems, totalPrice: totalPrice).boolValue }
    }

    func getOrderById(_ orderId: String) async -> Order? {
        return await run(fallback: nil as Order?) { try await sharedViewModel.getOrderById(orderId: orderId) }
    }

    func updateOrderStatus(_ orderId: String, status: String) async -> Bool {
        return await run(fallback: false) { try await sharedViewModel.updateOrderStatus(orderId: orderId, status: status).boolValue }
    }

    func cancelOrder(_ orderId: String) async -> Bool {
        return await run(fallback: false) { try await sharedViewModel.cancelOrder(orderId: orderId).boolValue }
    }

    // MARK: - Synchronous Methods

    func getOrdersByStatus(_ status: String) -> [Order] {
        return sharedViewModel.getOrdersByStatus(status: status)
    }

    func getTotalSpent() -> Double {
        return sharedViewModel.getTotalSpent()
    }

    func getOrderHistory(limit: Int = 10) -> [Order] {
        return sharedViewModel.getOrderHistory(limit: Int32(limit))
    }

    func getOrdersByDateRange(startDate: String, endDate: String) -> [Order] {
        return sharedViewModel.getOrdersByDateRange(startDate: startDate, endDate: endDate)
    }

    func canCancelOrder(_ orderId: String) -> Bool {
        return sharedViewModel.canCancelOrder(orderId: orderId)
    }

    func getOrderStatusDisplay(_ status: String) -> String {
        return sharedViewModel.getOrderStatusDisplay(status: status)
    }

    func getEstimatedDeliveryTime(_ orderId: String) -> String {
        return sharedViewModel.getEstimatedDeliveryTime(orderId: orderId)
    }

    func clearError() {
        sharedViewModel.clearError()
    }

    func refresh() {
        sharedViewModel.refresh()
    }

    // MARK: - Helper Methods for SwiftUI

    func formatPrice(_ price: Double) -> String {
        return price.asPrice
    }

    func formatTotalSpent() -> String {
        return formatPrice(state.totalSpent)
    }

    func getOrderStatusColor(_ status: String) -> Color {
        switch status.lowercased() {
        case "pending":
            return .orange
        case "processing":
            return .blue
        case "shipped":
            return .purple
        case "delivered":
            return .green
        case "cancelled":
            return .red
        default:
            return .gray
        }
    }

    func getOrderStatusIcon(_ status: String) -> String {
        switch status.lowercased() {
        case "pending":
            return "clock"
        case "processing":
            return "gear"
        case "shipped":
            return "shippingbox"
        case "delivered":
            return "checkmark.circle"
        case "cancelled":
            return "xmark.circle"
        default:
            return "questionmark.circle"
        }
    }

    func getOrdersGroupedByStatus() -> [String: [Order]] {
        return Dictionary(grouping: state.orders) { $0.status }
    }

    func getAverageOrderValue() -> Double {
        return sharedViewModel.getAverageOrderValue()
    }

    func formatAverageOrderValue() -> String {
        return formatPrice(getAverageOrderValue())
    }

    func getMostRecentOrder() -> Order? {
        return state.orders.first
    }

    func getOrdersThisMonth() -> [Order] {
        return sharedViewModel.getOrdersThisMonth()
    }
}
